oracle.go

```diff
@@ -0,0 +1,419 @@
+	// completed transaction.
+	CompletedTokens(tokens []string) (map[string]bool, error)
+
+	// RemoveTxns can be used to flag that a given transaction should not
+	// be considered part of the valid set.
+	RemoveTxns(txnIds []bson.ObjectId) error
+
+	// IterTxns lets you iterate over all of the transactions that have
+	// not been removed.
+	IterTxns() (OracleIterator, error)
+}
+
+// checkMongoSupportsOut verifies that Mongo supports "$out" in an aggregation
+// pipeline. This was introduced in Mongo 2.6
+// https://docs.mongodb.com/manual/reference/operator/aggregation/out/
+func checkMongoSupportsOut(db *mgo.Database) bool {
+	var dbInfo struct {
+		VersionArray []int `bson:"versionArray"`
+	}
+	if err := db.Run(bson.M{"buildInfo": 1}, &dbInfo); err != nil {
+		return false
+	}
+	logger.Debugf("buildInfo reported: %v", dbInfo.VersionArray)
+	if len(dbInfo.VersionArray) < 2 {
+		return false
+	}
+	// Check if we are at least 2.6
+	v := dbInfo.VersionArray
+	return v[0] > 2 || (v[0] == 2 && v[1] >= 6)
+}
+
+// NewDBOracle uses a database collection to manage the queue of remaining
+// transactions.
+// The caller is responsible to call the returned cleanup() function, to ensure
+// that any resources are freed.
+func NewDBOracle(db *mgo.Database, txns *mgo.Collection) (*DBOracle, func(), error) {
+	oracle := &DBOracle{
+		db:            db,
+		txns:          txns,
+		usingMongoOut: checkMongoSupportsOut(db),
+	}
+	cleanup, err := oracle.prepare()
+	return oracle, cleanup, err
+}
+
+var _ Oracle = (*DBOracle)(nil)
+
+func noopCleanup() {}
+
+// DBOracle uses a temporary table on disk to track what transactions are
+// considered completed and purgeable.
+type DBOracle struct {
+	db              *mgo.Database
+	txns            *mgo.Collection
+	working         *mgo.Collection
+	usingMongoOut   bool
+	checkedTokens   uint64
+	completedTokens uint64
+	foundTxns       uint64
+}
+
+// prepareWorkingDirectly iterates the working set from the pipeline and
+// populates the working set by inserting them from the client. This is less
+// efficient that a $out in the pipeline, but must be used when Mongo doesn't
+// support pipelines.
+func (o *DBOracle) prepareWorkingDirectly(pipeline []bson.M) error {
+	logger.Debugf("iterating the transactions collection to build the working set: %q", o.working.Name)
+	// Make sure the working set is clean
+	o.working.DropCollection()
+	pipe := o.txns.Pipe(pipeline)
+	pipe.Batch(maxBatchDocs)
+	pipe.AllowDiskUse()
+	iter := pipe.Iter()
+	var txnDoc struct {
+		Id bson.ObjectId `bson:"_id"`
+	}
+	t := newSimpleTimer(logInterval)
+	docCount := 0
+	// Batching the insert into 1000 at a time made a dramatic improvement in
+	// time. Doing one-by-one insert after 1hr wall-clock time it had only
+	// copied 11M transaction ids.
+	// With a 1000 item batch, it took 20min to do copy 36M documents (approx
+	// 10x speedup)
+	// For reference, it is about 9min to use $out, and 13min to read the data
+	// into memory.
```

Comment:
I didn't try it on large amounts of data with 2.4. I did run the test suite on 2.4. I don't have a large dump of 2.4 data accessible. Do you?
I did this specifically vs using the Bulk api because I expected insertion of many objects in a single pass to be better than one-by-one regardless. (less round-trips to mongo, it can lookup indexes and apply them to everything at once, etc.)
I'm reasonably confident that this will always be at least no worse than doing it one-by-one.